lib/crypto/c_src/api_ng.c

```diff
@@ -0,0 +1,155 @@
+/*
+ * %CopyrightBegin%
+ *
+ * Copyright Ericsson AB 2010-2018. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * %CopyrightEnd%
+ */
+
+#include "api_ng.h"
+#include "aes.h"
+#include "cipher.h"
+
+/*
+ * A unified set of functions for encryption/decryption.
+ *
+ * EXPERIMENTAL!!
+ *
+ */
+
+/* Try better error messages in new functions */
+#define ERROR_Term(Env, ReasonTerm) enif_make_tuple2(Env, atom_error, (ReasonTerm))
+#define ERROR_Str(Env, ReasonString) ERROR_Term(Env, enif_make_string(Env,ReasonString,ERL_NIF_LATIN1))
+
+/* Initializes state for (de)encryption
+ */
+ERL_NIF_TERM ng_crypto_init_nif(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
+{/* (Cipher, Key, IVec, Encrypt)  % if no IV for the Cipher, set IVec = <<>>
+ */
+    ErlNifBinary     key_bin, ivec_bin;
+    unsigned char *iv = NULL;
+    struct evp_cipher_ctx *ctx;
+    struct cipher_type_t *cipherp = NULL;
+    const EVP_CIPHER *cipher;
+    ERL_NIF_TERM ret;
+    int enc = (argv[argc-1] == atom_true)      ? 1  :
+              (argv[argc-1] == atom_false)     ? 0  :
+              (argv[argc-1] == atom_undefined) ? -1 : /* For compat funcs in crypto.erl */
+             -100,
+        iv_len = 0;
+
+    if (enc == -100)
+        return ERROR_Str(env, "Bad enc flag");
+
+    if (!enif_inspect_binary(env, argv[1], &key_bin))
+        return ERROR_Str(env, "Bad key");
+
+    if (!(cipherp = get_cipher_type(argv[0], key_bin.size)))
+        return ERROR_Str(env, "Unknown cipher or bad key size");
+
+    if (FORBIDDEN_IN_FIPS(cipherp))
+        return enif_raise_exception(env, atom_notsup);
+
+    if (!(cipher = cipherp->cipher.p)) {
+#if !defined(HAVE_EVP_AES_CTR)
+        if (cipherp->flags & AES_CTR_COMPAT)
+            return aes_ctr_stream_init_compat(env, argv[1], argv[2]);
+        else
+#endif
+            return enif_raise_exception(env, atom_notsup);
+    }
+
+    iv_len = EVP_CIPHER_iv_length(cipher);
+
+    if (iv_len) {
+        if (   !enif_inspect_binary(env, argv[2], &ivec_bin)
+            || (iv_len != ivec_bin.size)
+           ) {
+            return ERROR_Str(env, "Bad iv size");
+        }
+        iv = ivec_bin.data;
+    }
+
+    ctx = enif_alloc_resource(evp_cipher_ctx_rtype, sizeof(struct evp_cipher_ctx));
```

Comment:
Just a comment: `enif_alloc_resource` never returns NULL currently, but it seems like it should if I pass in something it can't allocate on my system like `SIZE_MAX`.